python/ray/data/dataset.py

```diff
@@ -697,21 +697,28 @@ def _map_batches_without_batch_size_validation(
     def add_column(
         self,
         col: str,
-        fn: Callable[["pandas.DataFrame"], "pandas.Series"],
+        fn: Union[
+            Callable[["pandas.DataFrame"], "pandas.Series"],
+            Callable[["pyarrow.Table"], "pyarrow.Array"],
```

Comment:
These API has to be consistent with `map_batches` for both inputs and outputs

Reply:
It wasn't consistent to begin with.
https://github.com/ray-project/ray/blob/f6439e1c4b8808686ef98465e7372f77a7859fdd/python/ray/data/dataset.py#L700

What should the type of `fn` be?

Reply:
What would the return type of the callable be? Currently it's `pandas.Series`. Changing the return type from `Series` to `DataFrame` would be a breaking change

Reply:
Map batches accepts "table"-like structure, here we expect list of column values so it's not to replace with DataFrame literally, but instead align the APIs

 - `map_batches` accepts: `DataBatch = Union["pyarrow.Table", "pandas.DataFrame", Dict[str, np.ndarray]]`
 - `add_column` should accept `Union[pa.Array, pandas.Seriers, ndarray]`

Reply:
> (ie also permit ndarray to be accepted/returned)

Oh, sorry. I missed this part. 

To clarify, is this what you had in mind? If so, makes sense to me for consistency with `map_batches`.
```python
Callable[[DataBatch], Union[pa.Array, pd.Series, ndarray]]
```

Reply:
> `Callable[[DataBatch], Union[pa.Array, pd.Series, ndarray]]`

@bveeramani That would not be correct because that term would allow for example `Callable[[pyarrow.Table], ndarray]]` which I don't think we want to allow?

Reply:
@alexeykudinkin Do you want us to add the numpy functionality in this PR as well for consistency with map_batches?

Reply:
Oh, I meant more from like an interface perspective. At least personally, I found it unexpected that I could do an Arrow table as input and a DataFrame as output (not that it's necessarily an issue)

Reply:
Thanks for the input guys. I've made the change. Waiting for CI...